### ISSUE TITLE
LVM + LUKS cryptsetup broken for dracut-ng-102 and 103

### ISSUE THREAD

**Issue by strenker** [dracut-ng/dracut-ng#563]
I have physical and virtual hosts using LVM with LUKS cryptsetup setup:
 1. LUKS / cryptsetup partition
 2. LVM based root and swap volumes on above LUKS partition

Gentoo Linux as distro with systemd as init-system and systemd-boot as boot-manager.

After migrating to dracut-ng-102 (from dracut-ng-60) all these systems refuse to boot. On the KVM/Qemu VM hosts i am able to capture the following error:

[ 1.239849] dracut-initqueue[418]: Failed to start systemd-cryptsetup@luks\x2dcde09b75\x2dd8f7\x2d4be0\x2d88ff\x2deb9c7f36327f.service: Unit systemd-cryptsetup@luks\x2dcde09b75\x2dd8f7\x2d4be0\x2d88ff\x2deb9c7f36327f.service not found.

I can reproduce the error with dracut-ng-102 and 103.
Dracut-ng-101 works as expected.

All the above with unchanged config-files.

The above error appears regardless if use dracut to generate a standard initrd-file or an signed UKI for direct UEFI-secure-boot.

The attached configs and logs are from my laptop which makes use of  the signed UKI and UEFI-secure-boot.

[blkid.txt](https://github.com/user-attachments/files/16491150/blkid.txt)
[dracut.conf.txt](https://github.com/user-attachments/files/16491151/dracut.conf.txt)
[install.conf.txt](https://github.com/user-attachments/files/16491152/install.conf.txt)
[lsblk-f.txt](https://github.com/user-attachments/files/16491153/lsblk-f.txt)
[make-install-dracut-101.txt](https://github.com/user-attachments/files/16491154/make-install-dracut-101.txt)
[make-install-dracut-102.txt](https://github.com/user-attachments/files/16491155/make-install-dracut-102.txt)

Maybe this issue is related to [issue](https://github.com/dracut-ng/dracut-ng/issues/451).

**Comment by ferringb** [comment]
A simple proposed fix here is to tweak https://github.com/dracut-ng/dracut-ng/commit/649e37bcd5ac46d5ac440b80297370c07ee7e1a8#diff-21b628d9fbf356aea5880a99b83b460c057ae0f064df37a58ca82c2cd8b34830L34-L36 and add a backwards compatibility fix like
```
if dracut_module_included systemd; then
  deps+=" systemd-cryptsetup "
  # add whatever warning you want here informing people that they need to update the invocation, or... arguable, just use crypt since this makes more sense anyways.
fi
```

**Comment by jozzsi** [comment]
Using 

```
add_dracutmodules+=" ${DRACUTMODULES[@]} "
```

instead of 
```
dracutmodules+=" ${DRACUTMODULES[@]} "
```

is preferred to avoid surprises like this one or even better drop the list and let dracut figure out a list (and have an omit list of you'd like to optimize). 

You can use the built-in dependency resolution to keep the list short and easy to reason about.

Instead of

```
DRACUTMODULES=(
	base
	crypt
	dracut-systemd
	fs-lib
	i18n
	kernel-modules
	kernel-modules-extra
	lvm
	qemu
	resume
	rootfs-block
	shutdown
	systemd
	systemd-initrd
	terminfo
	udev-rules
	usrmount
)
```

use


```
	systemd-cryptsetup
 	i18n
	kernel-modules
	kernel-modules-extra
	lvm
	qemu
	resume
	shutdown
	terminfo
	usrmount
```

You can probably also drop the `usrmount` module if you're using `systemd` .

**Comment by strenker** [comment]
I can confirm that adding `systemd-cryptsetup` to the dracut module list fixes the issue and is a feasible workaround. I have tested this workaround with dracut-ng-102 and 103. It is good for both.

Nevertheless i expect dracut to figure out by itself, which `systemd`-modules it requires to unlock an encrypted disk as it did until dracut-ng-101. To leave this with the user by surprise is not a good idea.

**Comment by floppym** [comment]
This seems like misconfiguration to me. I think @jozzsi has identified the issue here: don't use the dracutmodules option unless you are willing to update it in response to changes in dracut.

**Comment by floppym** [comment]
Actually, I see that the hostonly option is enabled here, which I would expect to cause systemd-cryptsetup to be included. Please disregard my previous comment.

**Comment by floppym** [comment]
I suppose the issue could be that the check() function for systemd-cryptsetup isn't even evaluated because of its absence in dracutmodules.

**Comment by strenker** [comment]
@floppym @ferringb  @jozzsi Indeed it is a misconfiguration on my side. When I started using dracut before around 3 years i was happy to create a configuration which booted my laptop. I googled a lot and was not aware of the subtleties of dracuts configuration elements and differences between `add_something` and `something`. In the last 24 hours i learned more about dracut configuration than in the 3 years before. _Adding_ modules with `add_dracutmodules` is sufficient and `systemd-cryptsetup` must not be added explicitly.

I apologize for all the mess i caused.

**Comment by floppym** [comment]
No worries; dracut is a beast and very few people fully understand how it works (myself included).

**Comment by ferringb** [comment]
Just noting; some of us use `dracutmodules` for a reason.  I switched to it not that long ago because no amount of fighting w/ add/omit could produce a UKI in the range I was happy about, and w/ things restricted as I wanted.  So yeah, I took responsibility on myself to manage that.

I think it's a reasonable ask that changes of this sort inform folks that compatibility is being broken for a module like this.  Performance implies optimization, which is pretty much how I missed it when I read the notes originally.

...however, if there was "breaking changes" in there, that's a different story.

**Comment by jozzsi** [comment]
@ferringb what do you think of the idea of providing a fix (even for the `dracutmodules` config) instead of warning the user ? 

https://github.com/dracut-ng/dracut-ng/pull/566 should include `systemd-cryptsetup` (as long as `dracut-systemd` and `crypt` gets included) .

**Comment by ferringb** [comment]
@jozzsi When I looked at the PR, it looks like it'll work if this issue is viewed as a regression.  However- and remember I'm not a maintainer- I think you're doing something the maintainers have avoided for modules.  This is the usage of conditional depends in modules:
``
ferringb@hrair ~/dracut-ng/modules.d $ for x in */module-setup.sh; do ( . "$x"; y=$(declare -f depends); [ "${y/dracut_module_include}" != "$y" ] && echo -e "${x}\n${y}\n" ); done
00shell-interpreter/module-setup.sh
depends () 
{ 
    shells='bash dash busybox';
    for shell in $shells;
    do
        if dracut_module_included "$shell"; then
            echo "$shell";
            return 0;
        fi;
    done;
    shell=$(realpath /bin/sh);
    shell=${shell##*/};
    echo "$shell";
    return 0
}

09dbus/module-setup.sh
depends () 
{ 
    local _module;
    for _module in dbus-daemon dbus-broker;
    do
        if dracut_module_included "$_module"; then
            echo "$_module";
            return 0;
        fi;
    done;
    if check_module "dbus-broker"; then
        echo "dbus-broker";
        return 0;
    fi;
    echo "dbus-daemon";
    return 0
}

40network/module-setup.sh
depends () 
{ 
    is_qemu_virtualized && echo -n "qemu-net ";
    for module in connman network-manager network-legacy systemd-networkd;
    do
        if dracut_module_included "$module"; then
            network_handler="$module";
            break;
        fi;
    done;
    if [ -z "$network_handler" ]; then
        if check_module "network-manager"; then
            network_handler="network-manager";
        else
            if check_module "systemd-networkd"; then
                network_handler="systemd-networkd";
            else
                if check_module "connman"; then
                    network_handler="connman";
                else
                    network_handler="network-legacy";
                fi;
            fi;
        fi;
    fi;
    echo "net-lib kernel-network-modules $network_handler";
    return 0
}
```
^^^ all of those are wiring *reverse dependencies*- the depend of that module says "I need this other one".  You're doing a forward dependency- "include systemd-cryptsetup if crypt is also on".  If a depend can't be satisfied- for whatever reason- my tracing of dracut code is that it basically takes out that entire chain.  Don't quote me on the last one, but I recall past "flex dracut and see what it does" behaving that way.  If so- and systemd-cryptsetup can't be included for some reason- that would take systemd out of the included modules.

I don't really have an answer here; I think the ontology setup for the modules is past what makes sense, and something like meta-modules (that just pull in others- like what you're trying to do) is a better approach here.  If crypt is supposed to just be the cryptsetup tools- call it that.  Leave crypt as something vaguely meta that pulls in the relevant crypt tools depending on other configured modules.

That approach can be it's own rats nest, but if you trace the install()'s of the other modules- this *is* how they do it now.  In fact, that's how it was before they broke systemd-cryptsetup out: https://github.com/dracut-ng/dracut-ng/commit/649e37bcd5ac46d5ac440b80297370c07ee7e1a8#diff-21b628d9fbf356aea5880a99b83b460c057ae0f064df37a58ca82c2cd8b34830L34-L36 .

Basically, that's my two cents; it's a decision for the maintainers what they want to allow w/in modules.  I'm just pointing out you're doing something that breaks from that pattern, although arguable, so did they.
